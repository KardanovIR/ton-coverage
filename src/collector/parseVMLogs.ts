export type LogEntry = {
    kind: 'execute',
    command: string
} | {
    kind: 'cell',
    hash: string,
    offset: number
} | {
    kind: 'stack',
    stack: string
} | {
    kind: 'gas',
    remaining: bigint
} | {
    kind: 'output_action'
}

export function parseVMLogs(logs: string) {
    let lines = logs.split('\n');
    let res: LogEntry[] = [];
    for (let l of lines) {
        l = l.trim();
        if (l.startsWith('gas remaining: ')) {
            let remaining = BigInt(l.substring(15));
            res.push({ kind: 'gas', remaining });
        } else if (l.startsWith('execute ')) {
            let command = l.substring(8);
            res.push({ kind: 'execute', command });
        } else if (l.startsWith('code cell hash: ')) {
            l = l.substring(16);
            let hash = l.substring(0, 64);
            let offset = parseInt(l.substring(65 + 'offset: '.length));
            res.push({ kind: 'cell', hash, offset: offset });
        } else if (l.startsWith('stack: ')) {
            let stack = l.substring(7);
            res.push({ kind: 'stack', stack });
        } else if (l.startsWith('installing an output action')) {
            res.push({ kind: 'output_action' });
        } else {
            throw new Error('Unknown log line: ' + l);
        }
    }
    return res;
}