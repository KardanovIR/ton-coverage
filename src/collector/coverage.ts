import { LogEntry } from "./parseVMLogs";

type CollectedCell = {
    offsets: Map<number, { count: number, gas: bigint[] }>;
}

export class CoverageCollector {
    #collected = new Map<string, CollectedCell>();

    coverageForCell(hash: string) {
        return this.#collected.get(hash);
    }

    collectCell(hash: Buffer, offset: number, gas: bigint) {
        let id = hash.toString('hex');
        let c = this.#collected.get(id);
        if (!c) {
            c = { offsets: new Map() };
            this.#collected.set(id, c);
        }
        let cc = c.offsets.get(offset);
        if (!cc) {
            cc = { count: 0, gas: [] };
            c.offsets.set(offset, cc);
        }
        cc.count++;
        cc.gas.push(gas);
    }

    export() {
        let res: any = {};
        for (let [id, c] of this.#collected) {
            let offsets: any = {};
            for (let [offset, cc] of c.offsets) {
                offsets[offset] = {
                    count: cc.count,
                    gas: cc.gas.map((v) => v.toString())
                }
            }
            res[id] = {
                offsets
            }
        }
        return res;
    }
}

export function collectCoverage(args: {
    collector: CoverageCollector,
    logs: LogEntry[],
    gasLimit: bigint
}) {
    let collector = args.collector;
    let logs = args.logs;
    let gasRemaining = args.gasLimit;
    let offset = 0;
    while (offset < logs.length) {

        // Load stack
        const stack = logs[offset++];
        if (stack.kind !== 'stack') {
            throw new Error('Expected stack log entry, got: ' + stack.kind + " at " + offset);
        }

        // Load cell
        let cell = logs[offset++];

        // Skip implicit RET
        if (cell.kind === 'execute' && cell.command.startsWith('implicit ')) {
            while (offset < logs.length) {
                if (logs[offset].kind === 'gas') {
                    break;
                }
                offset++;
            }
            let g = logs[offset++];
            if (g.kind !== 'gas') {
                throw new Error('Expected gas log entry, got: ' + g.kind + " at " + offset);
            }
            gasRemaining = g.remaining;
            continue;
        }

        // Check for cell
        if (cell.kind !== 'cell') {
            throw new Error('Expected cell log entry, got: ' + cell.kind + " at " + offset);
        }

        // Load execute
        let execute = logs[offset++];
        if (execute.kind !== 'execute') {
            throw new Error('Expected execute entry, got: ' + cell.kind + " at " + offset);
        }

        // Load gas
        let gas = logs[offset++];
        if (gas.kind === 'output_action') { // Skip output action
            gas = logs[offset++]
        }
        if (gas.kind !== 'gas') {
            throw new Error('Expected gas entry, got: ' + cell.kind + " at " + offset);
        }

        // Collect
        collector.collectCell(Buffer.from(cell.hash, 'hex'), cell.offset, gasRemaining - gas.remaining);

        // Update 
        gasRemaining = gas.remaining;
    }
}