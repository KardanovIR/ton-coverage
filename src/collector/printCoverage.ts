import { decompileAll, opcodeToString, Printer } from "@tact-lang/opcode";
import { Cell } from "ton-core";
import { pad, trimIndent } from "../utils/text";
import { CoverageCollector } from "./coverage";

function span(src: string, className: string) {
    return `<span class="${className}">${src}</span>`;
}

function template(src: string) {
    return trimIndent(`
    <html>
        <head>
            <style>
                pre {
                    display: flex;
                    flex-direction: column;
                }
                .padding {
                    color: #ccc;
                }
                .line-covered {
                    height: 1em;
                    background-color: #0f0;
                }
                .line-uncovered {
                    height: 1em;
                    background-color: #F6C6CE;
                }
                .count {
                    color: #00f;
                }
            </style>
        </head>
        <body>
            <pre>
            ${pad(trimIndent(src), 8)}
            </pre>
        <body>
    </html>
    `);
}

export function printCoverage(code: Cell, collector: CoverageCollector) {

    // Create printer
    const printer: Printer = (src, indent) => {
        if (typeof src === 'string') {
            return span(span('.'.repeat(indent * 2), 'padding') + src, 'line');
        }
        let line = span('.'.repeat(indent * 2), 'padding') + span(opcodeToString(src.op), 'opcode');
        let cov = collector.coverageForCell(src.hash);
        let lineClass: string;
        if (cov && cov.offsets.has(src.offset)) {
            let d = cov.offsets.get(src.offset)!!;
            line = line + ' ' + span(`x${d.count}`, 'count');
            let min = d.gas.reduce((a, b) => a < b ? a : b, d.gas[0]);
            let max = d.gas.reduce((a, b) => a > b ? a : b, d.gas[0]);
            if (min !== max) {
                line = line + ' ' + span(`gas: ${min} - ${max}`, 'gas');
            } else {
                line = line + ' ' + span(`gas: ${min}`, 'gas');
            }
            lineClass = 'line-covered';
        } else {
            lineClass = 'line-uncovered';
        }

        return span(line, lineClass);
    }

    // Decompile with coverage
    let decompiled = decompileAll({ src: code, printer });

    // Process template and return result
    return template(decompiled);
}