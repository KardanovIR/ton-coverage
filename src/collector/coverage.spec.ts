import * as fs from 'fs';
import * as path from 'path';
import { Cell } from 'ton-core';
import { collectCoverage, CoverageCollector } from './coverage';
import { parseVMLogs } from './parseVMLogs';
import { printCoverage } from './printCoverage';

describe('coverage', () => {
    it('should parse logs', () => {
        let logs = fs.readFileSync(path.resolve(__dirname, '__testdata__', 'log1.txt'), 'utf8');
        parseVMLogs(logs);
    });
    it('should collect coverage', () => {

        // Collect coverage
        let rawLogs = fs.readFileSync(path.resolve(__dirname, '__testdata__', 'log1.txt'), 'utf8');
        let logs = parseVMLogs(rawLogs);
        let collector = new CoverageCollector();
        collectCoverage({
            collector,
            logs,
            gasLimit: 1000000000n
        });
        expect(collector.export()).toMatchSnapshot();

        // Print coverage
        let rawCode = Cell.fromBoc(fs.readFileSync(path.resolve(__dirname, '__testdata__', 'log1.boc')))[0];
        let res = printCoverage(rawCode, collector);
        fs.writeFileSync(path.resolve(__dirname, '__testdata__', 'log1.html'), res);
    });
});