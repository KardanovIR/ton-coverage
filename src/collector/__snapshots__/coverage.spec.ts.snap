// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coverage should collect coverage 1`] = `
{
  "208b79f9ac4a90e87893688d1f6526547fb83f687f624343063f557865582c60": {
    "offsets": {
      "104": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "112": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "120": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "128": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "144": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "152": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "16": {
        "count": 1,
        "gas": [
          "118",
        ],
      },
      "160": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "176": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "200": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "208": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "232": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "240": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "264": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "272": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "288": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "296": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "304": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "328": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "336": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "344": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "368": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "392": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "400": {
        "count": 1,
        "gas": [
          "30",
        ],
      },
      "416": {
        "count": 1,
        "gas": [
          "28",
        ],
      },
      "432": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "440": {
        "count": 1,
        "gas": [
          "29",
        ],
      },
      "48": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "56": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "64": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "8": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "96": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
    },
  },
  "2b98264ed7626f5e2efb0798866f46d1bd3861c759972fff82693576e5bb7199": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "104": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "112": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "128": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "152": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "160": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "184": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "192": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "216": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "224": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "240": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "248": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "256": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "280": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "288": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "296": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "32": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "320": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "336": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "344": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "360": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "368": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "384": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "400": {
        "count": 1,
        "gas": [
          "518",
        ],
      },
      "408": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "48": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "56": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "72": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "8": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "88": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
    },
  },
  "4c95df20e52b520371a8124d274abaaee5652c415480ccd15de3c6f1cd6243e6": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "16": {
        "count": 1,
        "gas": [
          "118",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "32": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "30",
        ],
      },
      "424": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "440": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "456": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "464": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "472": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "488": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "512": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "520": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "544": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "552": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "56": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "576": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "584": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "600": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "608": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "616": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "640": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "648": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "656": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "680": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "688": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "712": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "72": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "728": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "752": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "768": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "784": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "792": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "800": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "808": {
        "count": 1,
        "gas": [
          "126",
        ],
      },
      "824": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "832": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "848": {
        "count": 1,
        "gas": [
          "126",
        ],
      },
      "864": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
    },
  },
  "644ef5250652bdfa6844e3e13a61c1eb6a52688e5503308f9a5359c7c6dbdb76": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
    },
  },
  "90215e20bdae5e1e62cfe50feabe14bf0c514e00468d34de3606af250b13660a": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "104": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "128": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "136": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "144": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "16": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "168": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "176": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "184": {
        "count": 1,
        "gas": [
          "23",
        ],
      },
      "232": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "240": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "32": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "48": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "64": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "672": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "680": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "688": {
        "count": 1,
        "gas": [
          "23",
        ],
      },
      "736": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "744": {
        "count": 1,
        "gas": [
          "126",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "88": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "96": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
    },
  },
  "b313972e62482409709260dedba7be6eb84de7b9ee9e9c2794517e490d57349e": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "112": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "128": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "136": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "152": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "16": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "176": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "184": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "208": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "216": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "240": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "248": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "264": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "272": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "280": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "304": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "312": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "320": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "344": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "360": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "376": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "392": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "400": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "408": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "424": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "432": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "440": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "448": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "464": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "472": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "480": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "488": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "496": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "504": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "512": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "520": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "528": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "536": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "544": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "552": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "56": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "568": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "584": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "592": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "600": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "608": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "616": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "624": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "632": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "64": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "648": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "656": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "664": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "672": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "696": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "704": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "712": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "760": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "768": {
        "count": 1,
        "gas": [
          "518",
        ],
      },
      "776": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "784": {
        "count": 1,
        "gas": [
          "526",
        ],
      },
      "8": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "88": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "96": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
    },
  },
  "b7bc280e2c801e7cd5292a213058754302b93c37a7b149c5ba4b0309a9573108": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "32": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "48": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "64": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "72": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "8": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "88": {
        "count": 1,
        "gas": [
          "126",
        ],
      },
    },
  },
  "c68f86303a9e9f313486e75cbeee7072aa8f350b3e5eb1923e641066fed416ca": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "112": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "128": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "136": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "144": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "152": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "160": {
        "count": 1,
        "gas": [
          "23",
        ],
      },
      "208": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "216": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "232": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "24": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "248": {
        "count": 1,
        "gas": [
          "518",
        ],
      },
      "256": {
        "count": 1,
        "gas": [
          "126",
        ],
      },
      "272": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "32": {
        "count": 1,
        "gas": [
          "23",
        ],
      },
      "8": {
        "count": 1,
        "gas": [
          "26",
        ],
      },
      "80": {
        "count": 1,
        "gas": [
          "18",
        ],
      },
      "88": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
    },
  },
  "fb2da23fc4410cd7da1cefbcf8b9cf088f45ad32a418de8bd1c72354c60c99c9": {
    "offsets": {
      "0": {
        "count": 1,
        "gas": [
          "999000026",
        ],
      },
      "16": {
        "count": 1,
        "gas": [
          "34",
        ],
      },
      "40": {
        "count": 1,
        "gas": [
          "226",
        ],
      },
    },
  },
}
`;
